File: examples-copy/playground/events/subtask-service.py
import os
from typing import Optional

from .event_bus import IEventBus
from .repositories import TaskRepository
from .type_defs import (
    CompleteSubtaskCommand,
    NextSubtaskTriggeredEvent,
    StartNewChatCommand,
    StartSubtaskCommand,
    SubtaskCompletedEvent,
    SubtaskStartedEvent,
    SubtaskUpdatedEvent,
)


class SubtaskService:
    def __init__(self, event_bus: IEventBus, task_repo: TaskRepository):
        self._event_bus = event_bus
        self._task_repo = task_repo

        # Subscribe to events
        self._event_bus.subscribe(
            'StartSubtaskCommand', self._handle_start_subtask_command
        )
        self._event_bus.subscribe(
            'CompleteSubtaskCommand', self._handle_complete_subtask_command
        )
        self._event_bus.subscribe('SubtaskUpdatedEvent', self._on_subtask_updated)

    async def _handle_start_subtask_command(self, command: StartSubtaskCommand) -> None:
        """Handle the start subtask command and manage the task execution flow.

        Steps:
        1. Get task item and subtask item from store
        2. Get subtask input from previous output or initial input
        3. Check running task state and handle accordingly
        4. Create/verify subtask folder
        5. Start new chat and emit events
        """
        task, subtask = self._task_repo.get_subtask(command.task_id, command.subtask_id)

        # Get subtask's input
        input_data = await self._get_subtask_input(command.subtask_id)

        # Handle running task state
        if task.status == 'IN_PROGRESS' and task.current_subtask_id:
            if task.current_subtask_id == command.subtask_id:
                print(f'Warning: Subtask {command.subtask_id} is already running')
                return

            # Pause currently running subtask
            await self._pause_subtask(command.task_id, command.subtask_id)

        # Check and create subtask folder if needed
        has_folder = await self._check_subtask_folder(
            command.task_id, command.subtask_id
        )
        if not has_folder:
            # Initialize subtask first
            await self._init_subtask(command.task_id, command.subtask_id)
            await self._create_subtask_folder(command.task_id, command.subtask_id)

        # Start new chat (fire and forget)
        await self._event_bus.publish(
            StartNewChatCommand(
                task_id=command.task_id,
                subtask_id=command.subtask_id,
                timestamp=command.timestamp,
            )
        )

        # Update subtask status
        await self._event_bus.publish(
            SubtaskUpdatedEvent(
                task_id=command.task_id,
                subtask_id=command.subtask_id,
                status='in_progress',
                timestamp=command.timestamp,
            )
        )

        # Mark subtask as started
        await self._event_bus.publish(
            SubtaskStartedEvent(
                task_id=command.task_id,
                subtask_id=command.subtask_id,
                input=input_data,
                timestamp=command.timestamp,
            )
        )

    async def _handle_complete_subtask_command(
        self, command: CompleteSubtaskCommand
    ) -> None:
        """Steps to complete a subtask.

        1. Move previous output to history
        2. Save new output
        3. Update subtask status
        4. Trigger next subtask
        """
        task, subtask = self._task_repo.get_subtask(command.task_id, command.subtask_id)

        # Handle output files
        subtask_path = self._get_subtask_folder_path(
            command.task_id, command.subtask_id
        )
        history_path = os.path.join(subtask_path, 'history')
        os.makedirs(history_path, exist_ok=True)

        # Move existing output to history with timestamp
        try:
            from datetime import datetime

            timestamp = datetime.now().strftime('%Y%m%d_%H%M%S')
            current_output_path = os.path.join(subtask_path, 'output.json')

            if os.path.exists(current_output_path):
                with open(current_output_path) as f:
                    current_output = f.read()

                history_file = os.path.join(history_path, f'output_{timestamp}.json')
                with open(history_file, 'w') as f:
                    f.write(current_output)
        except Exception:
            # Ignore if no existing output
            pass

        # Save new output
        with open(os.path.join(subtask_path, 'output.json'), 'w') as f:
            f.write(command.output)

        # Update subtask status and emit event
        await self._event_bus.publish(
            SubtaskUpdatedEvent(
                task_id=command.task_id,
                subtask_id=command.subtask_id,
                status='completed',
                timestamp=command.timestamp,
            )
        )

        # Complete subtask and trigger next
        await self._event_bus.publish(
            SubtaskCompletedEvent(
                task_id=command.task_id,
                subtask_id=command.subtask_id,
                timestamp=command.timestamp,
            )
        )

        await self._event_bus.publish(
            NextSubtaskTriggeredEvent(
                task_id=command.task_id,
                current_subtask_id=command.subtask_id,
                timestamp=command.timestamp,
            )
        )

    async def _on_subtask_updated(self, event: SubtaskUpdatedEvent) -> None:
        """Updates subtask status in repository when status changes."""
        task, subtask = self._task_repo.get_subtask(event.task_id, event.subtask_id)
        subtask.status = event.status
        self._task_repo.save_subtask(subtask)

    # ----------------
    # Helper Methods
    # ----------------

    async def _get_subtask_input(self, subtask_id: str) -> Optional[str]:
        """Gets input for a subtask from either specified sources.

        1. Previous subtask's output.json
        2. Initial input from task store for first subtask

        Note: Currently using string input/output only
        File structure:
            workspace/task_id/
                step1_validate_input/
                    output.json
                    history/
                        output_YYYYMMDD_HHMMSS.json
        """
        # TODO: Implement actual input retrieval logic
        return 'Example input string'

    def _get_subtask_folder_path(
        self, task_id: str, subtask_id: str, workspace: str = 'tasks'
    ) -> str:
        return os.path.join(workspace, task_id, 'subtasks', subtask_id)

    async def _check_subtask_folder(
        self, task_id: str, subtask_id: str, workspace: str = 'tasks'
    ) -> bool:
        folder_path = self._get_subtask_folder_path(task_id, subtask_id, workspace)
        return os.path.exists(folder_path)

    async def _init_subtask(self, task_id: str, subtask_id: str) -> None:
        raise NotImplementedError('Not yet implemented')

    async def _create_subtask_folder(self, task_id: str, subtask_id: str) -> None:
        raise NotImplementedError('Not yet implemented')

    async def _pause_subtask(self, task_id: str, subtask_id: str) -> None:
        """Handles pausing of a running subtask.

        Notes:
        1. Handles only subtask scope; task scope handled via SubtaskPaused event
        2. No parallel subtasks support yet
        3. Pausing subtask = pausing task for now

        Steps:
        - Verify subtask state and ID
        - Update task status
        - Save progress
        - Emit events
        """
        raise NotImplementedError('Not yet implemented')

File: examples-copy/docs/event-driven/events/event-system-ts/subtask-service.ts
import { IEventBus } from "./event-system";
import { TaskRepository } from "./repositories";
import {
  CompleteSubtaskCommand,
  NextSubtaskTriggeredEvent,
  StartNewChatCommand,
  StartSubtaskCommand,
  Subtask,
  SubtaskCompletedEvent,
  SubtaskStartedEvent,
  SubtaskUpdatedEvent,
  Task,
} from "./types";
import { promises as fs } from "fs";
import * as path from "path";

class SubtaskService {
  constructor(private eventBus: IEventBus, private taskRepo: TaskRepository) {
    this.eventBus.subscribe<StartSubtaskCommand>(
      "StartSubtaskCommand",
      this.handleStartSubtaskCommand.bind(this)
    );
    this.eventBus.subscribe<CompleteSubtaskCommand>(
      "CompleteSubtaskCommand",
      this.handleCompleteSubtaskCommand.bind(this)
    );
    this.eventBus.subscribe<SubtaskUpdatedEvent>(
      "SubtaskUpdatedEvent",
      this.onSubtaskUpdated.bind(this)
    );
  }

  /**
   * Steps
   * - Get the task item from the store. If not found, raise error
   * - Get the subtask item from the task item.
   * - Check is this subtask is prepared by
   *   - Need to have the input. If not, try to get it from the previous subtask's output.
   * - Check is the task currently running?
   *   - If yes, then check which subtask is running.
   *     - If the running subtask is this. Do nothing. Log a warnning message.
   *     - If the running subtask is not this. Pause the running subtask first.
   * - Check is the subtask folder is existed, if not, create one.
   *   - If is existed. Do nothing.
   * - emit 'StartNewChatCommand' -> This will trigger the proceeded process. No need to wait.
   * - emit 'SubtaskStarted'
   */
  private async handleStartSubtaskCommand(
    command: StartSubtaskCommand
  ): Promise<void> {
    const { taskId, subtaskId } = command;

    // Get task and subtask from repository
    const { task, subtask } = this.taskRepo.getSubtask(taskId, subtaskId);

    // Get subtask's input
    const input = await this.getSubtaskInput(subtaskId);

    // Handle running task state
    if (task.status === "IN_PROGRESS" && task.currentSubtaskId) {
      if (task.currentSubtaskId === subtaskId) {
        console.warn(`Subtask ${subtaskId} is already running`);
        return;
      }

      // Pause the currently running subtask
      await this.pauseSubtask(taskId, subtaskId);
    }

    // Check and create subtask folder if needed
    const hasFolder = await this.checkSubtaskFolder(taskId, subtaskId);
    if (!hasFolder) {
      // Initialize subtask first
      await this.initSubtask(taskId, subtaskId);

      // Create folder and emit event
      await this.createSubtaskFolder(taskId, subtaskId);
    }

    // Start new chat (fire and forget)
    this.eventBus.publish<StartNewChatCommand>({
      type: "StartNewChatCommand",
      taskId,
      subtaskId,
      // input,
      timestamp: Date.now(),
    });

    // Update subtask status
    await this.eventBus.publish<SubtaskUpdatedEvent>({
      type: "SubtaskUpdatedEvent",
      taskId,
      subtaskId,
      status: "in_progress",
      timestamp: Date.now(),
    });

    // Mark subtask as started
    await this.eventBus.publish<SubtaskStartedEvent>({
      type: "SubtaskStartedEvent",
      taskId,
      subtaskId,
      input,
      timestamp: Date.now(),
    });
  }

  /**
   * This can be triggered when 1. the user approves the subtask output, 2. the subtask output is auto-approved.
   *
   * Steps:
   * 1. Move the previous subtask's output to the history folder.
   * 2. Save the output. The file name should be 'output.json'.
   * 3. Update the subtask status. Emit 'SubtaskUpdated' event.
   * 4. Emit NextSubtaskTriggeredEvent.
   */
  private async handleCompleteSubtaskCommand(
    command: CompleteSubtaskCommand
  ): Promise<void> {
    const { taskId, subtaskId, output } = command;

    // 1. Get the task and subtask
    const { task, subtask } = this.taskRepo.getSubtask(taskId, subtaskId);

    // 2. Save the current output to history
    const subtaskPath = this.getSubtaskFolderPath(taskId, subtaskId);
    const historyPath = path.join(subtaskPath, "history");
    await fs.mkdir(historyPath, { recursive: true });

    // Move existing output to history with timestamp
    try {
      const timestamp = new Date().toISOString().replace(/[:.]/g, "");
      const currentOutput = await fs.readFile(
        path.join(subtaskPath, "output.json")
      );
      await fs.writeFile(
        path.join(historyPath, `output_${timestamp}.json`),
        currentOutput
      );
    } catch (error) {
      // Ignore if no existing output
    }

    // 3. Save new output
    await fs.writeFile(
      path.join(subtaskPath, "output.json"),
      JSON.stringify(output, null, 2)
    );

    // 4. Update subtask status and emit event
    await this.eventBus.publish<SubtaskUpdatedEvent>({
      type: "SubtaskUpdatedEvent",
      taskId,
      subtaskId,
      status: "completed",
      timestamp: Date.now(),
      // correlationId,
    });

    // 5. Complete subtask and trigger next
    await this.eventBus.publish<SubtaskCompletedEvent>({
      type: "SubtaskCompletedEvent",
      taskId,
      subtaskId,
      timestamp: Date.now(),
      // correlationId,
    });

    await this.eventBus.publish<NextSubtaskTriggeredEvent>({
      type: "NextSubtaskTriggeredEvent",
      taskId,
      currentSubtaskId: subtaskId,
      timestamp: Date.now(),
      // correlationId,
    });
  }

  private async onSubtaskUpdated(event: SubtaskUpdatedEvent): Promise<void> {
    const { subtask } = this.taskRepo.getSubtask(event.taskId, event.subtaskId);
    subtask.status = event.status;
    this.taskRepo.saveSubtask(subtask);
  }

  // Helper methods

  private async getSubtaskInput(subtaskId: string): Promise<string> {
    /**
     * Gets the input data for a subtask.
     *
     * The function retrieves input from one of the following sources, in priority order:
     * 1. Previous subtask's output.json file located at:
     *    ${workspace}/${task_id}/${prev_step_number}_${prev_subtask_name}/output.json
     *
     * 2. For the first subtask in a sequence:
     *    Initial input stored in TaskStore
     *
     * @remarks
     * - Currently assumes all input is of type 'string' with no parsing required
     * - File structure follows the convention:
     *   ${workspace}/${task_id}/
     *     step1_validate_input/
     *       output.json             # Latest output
     *       history/                # Historical records
     *         output_YYYYMMDD_HHMMSS.json
     *
     * @example Output JSON format:
     * {
     *   "summary": {
     *     "step": "驗證輸入資料",
     *     "status": "成功",
     *     "completed_at": "2024-01-13 15:30:00",
     *     "message": "驗證通過：共處理 100 筆資料"
     *   },
     *   "result": {
     *     // Actual output content
     *   }
     * }
     */

    // TODO
    return "This is an example output string";
  }

  private getSubtaskFolderPath(
    taskId: string,
    subtaskId: string,
    workspace: string = "tasks"
  ): string {
    return path.join(workspace, taskId, "subtasks", subtaskId);
  }

  private async checkSubtaskFolder(
    taskId: string,
    subtaskId: string,
    workspace: string = "tasks"
  ): Promise<boolean> {
    // Implementation to check if folder exists
    try {
      await fs.access(this.getSubtaskFolderPath(workspace, taskId, subtaskId));
      return true;
    } catch {
      return false;
    }
  }

  private async initSubtask(taskId: string, subtaskId: string): Promise<void> {
    throw new Error("Not yet implemented");
  }

  private async createSubtaskFolder(
    taskId: string,
    subtaskId: string
  ): Promise<void> {
    throw new Error("Not yet implemented");
  }

  private async pauseSubtask(taskId: string, subtaskId: string): Promise<void> {
    /**
     * Remarks:
     * 1. This method only consider the subtask's scope. For the task's scope, it will process the rest through 'SubtaskPaused' event. 
     * 2. For now we don't consider parallel subtasks, so when pausing the subtask, the task is also paused.
     * 3. Although we seperate the pauseSubtask and pauseTask methods, there is no difference for now.
     * 
     * Steps:
Subtask Status Check
Verify subtaskId matches current running subtask
Check if subtask is in a pausable state

Pause Actions
Update task status in store
Clear currentSubtaskId
Save any in-progress work

Event Publishing
Emit SubtaskPausedEvent
Include taskId, subtaskId, timestamp
Use consistent correlationId pattern
     */
    throw new Error("Not yet implemented");
  }
}
